### PR TITLE
Feature/PO4 - Model Genesis Block & Stake Distribution

### PR DESCRIPTION
**Pull request by javierdiaz72**
The intention of this PR is to provide an initial implementation of the genesis block class and the calculation of relative stakes given stake distributions and absolute stakes.

### CODE REVIEW

File: src/main/scala/io/iohk/praos/domain/RelativeStake.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    * Creates a relative stake given a stakeholder's public key and a given stake distribution. The relative stake is
+    * calculated as the stake (as indicated in the stake distribution) held by the particular stakeholder identified by
+    * the given public key, divided by the total stake held by all the stakeholders (as indicated in the stake
+    * distribution).
+    *
+    * @param publicKey The public key used to identify a particular stakeholder.
+    * @param stakeDistribution The stake distribution to be used in the calculations.
+    */
+  def apply(publicKey: Key, stakeDistribution: StakeDistribution): RelativeStake = {
+    val stakeholderStake = stakeDistribution(publicKey)
```
Comment (javierdiaz72) [inline src/main/scala/io/iohk/praos/domain/RelativeStake.scala]: Add "TODO: Check that `stakeDistribution` contains `publicKey` and issue an error if check fails, or add a  precondition".

File: src/main/scala/io/iohk/praos/domain/RelativeStake.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+    * Creates a relative stake given two absolute stakes: The total stake held by a particular stakeholder (denoted by
+    * si) and the total stake held by all the stakeholders (denoted by sP). The relative stake is calculated as si/sP.
+    *
+    * @param stakeholderStake The total stake held by a particular stakeholder.
+    * @param totalStake The total stake held by all the stakeholders.
+    */
+  def apply(stakeholderStake: Stake, totalStake: Stake): RelativeStake = {
+    RelativeStake(stakeholderStake.toFloat / totalStake.toFloat)
```
Comment (javierdiaz72) [inline src/main/scala/io/iohk/praos/domain/RelativeStake.scala]: Add "TODO: Check that `totalStake > 0 && totalStake >= stakeholderStake` and issue an error if check fails, or add a  precondition".

File: src/main/scala/io/iohk/praos/domain/package.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    * An amount of stake held by a stakeholder.
+    */
+  type Stake = Int
+
+  /**
+    * A stake distribution. As noted in Section 2.2.1 of Praos Formalization, stakeholders are identified by their
+    * public keys.
+    */
+  type StakeDistribution = Map[Key, Stake]
+  val StakeDistribution = Map
```
Comment (javierdiaz72) [inline src/main/scala/io/iohk/praos/domain/package.scala]: Check if this can be restricted.

File: src/main/main.iml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (javierdiaz72) [inline src/main/main.iml]: Add to .gitignore.

File: src/test/test.iml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (javierdiaz72) [inline src/test/test.iml]: Add to .gitignore.

File: src/main/scala/io/iohk/praos/domain/RelativeStake.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package io.iohk.praos.domain
+
+import io.iohk.praos.crypto.Key
+
+
+/**
+  * Factory for [[io.iohk.praos.domain.RelativeStake]] instances.
+  */
+object RelativeStake {
```
Comment (javierdiaz72) [inline src/main/scala/io/iohk/praos/domain/RelativeStake.scala]: Use a RelativeStakeCalculator or RelativeStakeFactory to create instances for RelativeStake and replace case class for a type alias.

File: src/main/main.iml (hunk 4 above)
Comment (javierdiaz72) [reply src/main/main.iml]: Resolved.

File: src/main/scala/io/iohk/praos/domain/RelativeStake.scala (hunk 6 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/RelativeStake.scala]: Resolved.

File: src/main/scala/io/iohk/praos/domain/RelativeStake.scala (hunk 1 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/RelativeStake.scala]: Resolved.

File: src/test/test.iml (hunk 5 above)
Comment (javierdiaz72) [reply src/test/test.iml]: Resolved.

File: src/main/scala/io/iohk/praos/domain/package.scala (hunk 3 above)
Comment (javierdiaz72) [reply src/main/scala/io/iohk/praos/domain/package.scala]: Resolved.

**Review by ntallar — APPROVED** [review]

**Pull request merged**